spec/config_spec.rb

```diff
@@ -665,6 +665,57 @@ def self.name
       end
     end
 
+    context "with variety of required attributes combinations with env option" do
+      before { allow(Anyway::Settings).to receive(:current_environment).and_return("production") }
+
+      let(:app_config) do
+        Class.new(described_class) do
+          config_name "cool_app"
+
+          attr_config :host, :port, :api_key, :user, :password, :debug, :redis_host
+          required :host, :port
+          required :api_key, env: "production"
+          required :user, :password, env: %w[development production]
+          required :redis_host, env: {except: :test}
+        end
+      end
+
+      let(:demo_config) do
+        Class.new(described_class) do
+          config_name "demo"
+          attr_config :sentry_api_key
+
+          required :sentry_api_key, env: {except: %i[local development test]}
+        end
+      end
+
+      it "raises ValidationError when required params without env option is missed" do
+        expect { app_config.new(api_key: "123", user: "john", password: "simple", redis_host: "localhost:6379") }
+          .to raise_error(Anyway::Config::ValidationError, /missing or empty: host, port/)
```

Comment:
Here and below: shouldn't be add `$` (or whatever) to the regexp to make sure that all the missing attributes are covered? I mean, `/missing or empty: host, port/` also matches "missing or empty: host, port, user, password".

Or, maybe, we can use a composed matcher to check for missing substrings. Not sure.

One approach could be to start with the valid attributes set as the base, and exclude required attributes one by one.